metricbeat/helper/kubernetes/state_metricset.go

```diff
@@ -0,0 +1,112 @@
+// under the License.
+
+package kubernetes
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/elastic/beats/v7/metricbeat/helper/prometheus"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	k8smod "github.com/elastic/beats/v7/metricbeat/module/kubernetes"
+	"github.com/elastic/beats/v7/metricbeat/module/kubernetes/util"
+)
+
+const prefix = "state_"
+
+var mappings = map[string]*prometheus.MetricsMapping{}
+
+// Init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func Init(name string, mapping *prometheus.MetricsMapping) {
+	name = prefix + name
+	mappings[name] = mapping
+	mb.Registry.MustAddMetricSet("kubernetes", name, New, mb.WithHostParser(prometheus.HostParser))
+}
+
+// MetricSet type defines all fields of the MetricSet
+// As a minimum it must inherit the mb.BaseMetricSet fields, but can be extended with
+// additional entries. These variables can be used to persist data or configuration between
+// multiple fetch calls.
+type MetricSet struct {
+	mb.BaseMetricSet
+	prometheusClient   prometheus.Prometheus
+	prometheusMappings *prometheus.MetricsMapping
+	mod                k8smod.Module
+	enricher           util.Enricher
+}
+
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	prometheusClient, err := prometheus.NewPrometheusClient(base)
+	if err != nil {
+		return nil, err
+	}
+	mod, ok := base.Module().(k8smod.Module)
+	if !ok {
+		return nil, fmt.Errorf("must be child of kubernetes module")
+	}
+
+	mapping := mappings[base.Name()]
+	var ms = MetricSet{
+		BaseMetricSet:      base,
+		prometheusClient:   prometheusClient,
+		prometheusMappings: mapping,
+		enricher:           util.NewResourceMetadataEnricher(base, strings.ReplaceAll(base.Name(), prefix, ""), mod.GetMetricsRepo(), false),
+		mod:                mod,
+	}
+	return &ms, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	m.enricher.Start()
+
+	families, err := m.mod.GetStateMetricsFamilies(m.prometheusClient)
+	if err != nil {
+		m.Logger().Error(err)
+		reporter.Error(err)
+		return
+	}
+	events, err := m.prometheusClient.ProcessMetrics(families, m.prometheusMappings)
+	if err != nil {
+		m.Logger().Error(err)
+		reporter.Error(err)
+		return
+	}
+
+	m.enricher.Enrich(events)
+	for _, event := range events {
+		e, err := util.CreateEvent(event, "kubernetes."+strings.ReplaceAll(m.BaseMetricSet.Name(), "state_", ""))
+		if err != nil {
+			m.Logger().Error(err)
+		}
+
+		if reported := reporter.Event(e); !reported {
+			m.Logger().Debug("error trying to emit event")
+			return
+		}
+	}
+}
+
+// Close stops this metricset
+func (m *MetricSet) Close() error {
+	m.enricher.Stop()
```

Comment:
Shouldn't be checked for `nil` as it was the case at https://github.com/elastic/beats/pull/34432/files#diff-dbfdc4ca83ad3d212a1f302ebc1bf34c8a123be72b14d0eeccaafb0455cc3146L129 for example?

Reply:
Not needed, because of this part of the [code](https://github.com/elastic/beats/blob/75071ccfb825c28003a560c4e889687944efaa7d/metricbeat/module/kubernetes/util/kubernetes.go#L577-L581).